src/boot/efi/pe.c

```diff
@@ -153,15 +155,32 @@ static void locate_sections(
         assert(offsets);
         assert(sizes);
 
+        EFI_STATUS err;
+
         for (size_t i = 0; i < n_table; i++) {
                 const PeSectionHeader *sect = section_table + i;
 
                 for (size_t j = 0; sections[j]; j++) {
                         if (memcmp(sect->Name, sections[j], strlen8(sections[j])) != 0)
                                 continue;
 
-                        offsets[j] = in_memory ? sect->VirtualAddress : sect->PointerToRawData;
-                        sizes[j] = sect->VirtualSize;
+                        size_t offset = in_memory ? sect->VirtualAddress : sect->PointerToRawData;
+                        size_t size = sect->VirtualSize;
+
+                        offsets[j] = offsets[j] ? : offset;
+                        sizes[j] = sizes[j] ? : size;
+
+                        if (in_memory && streq8(sections[j], ".dtb")) {
+                                err = devicetree_match(base + offset, size);
+
+                                if (err == EFI_SUCCESS) {
+                                        offsets[j] = offset;
+                                        sizes[j] = size;
+                                }
+
+                                if (err == EFI_INVALID_PARAMETER)
```

Comment:
why just handle this one error, instead of all that are not EFI_SUCCESS here?

Reply:
I singled it out since devicetree_match() returns EFI_UNSUPPORTED in the case where no FW dtb is found as opposed to EFI_INVALID_PARAMETER where the UKI dtbs are not valid